[Skia] Use Skia API when adding rounded rect to path and strategy is PreferNative

carlosgcampos opened the pull request:
#### 7d9a6549eb1269c0d823ac95fe940f17f49a4b53
<pre>
[Skia] Use Skia API when adding rounded rect to path and strategy is PreferNative
<a href="https://bugs.webkit.org/show_bug.cgi?id=269845">https://bugs.webkit.org/show_bug.cgi?id=269845</a>

Reviewed by Adrian Perez de Castro.

Also add conversion operators between FloatRounrdedRect and SkRREct to
remove duplicated code.

* Source/WebCore/platform/SourcesSkia.txt:
* Source/WebCore/platform/graphics/FloatRoundedRect.h:
* Source/WebCore/platform/graphics/skia/FloatRoundedRectSkia.cpp: Added.
(WebCore::FloatRoundedRect::FloatRoundedRect):
(WebCore::FloatRoundedRect::operator SkRRect const):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::fillRoundedRectImpl):
(WebCore::GraphicsContextSkia::fillRectWithRoundedHole):
* Source/WebCore/platform/graphics/skia/PathSkia.cpp:
(WebCore::PathSkia::add):

Canonical link: <a href="https://commits.webkit.org/275106@main">https://commits.webkit.org/275106@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/06df9c0a41bc8bc9b5adc7f6b1864ae5a41011db

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40893 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19906 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43271 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43451 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36984 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43200 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22910 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17237 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33894 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41467 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16853 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35252 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14517 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14622 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36242 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44746 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37107 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36553 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40304 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15708 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12899 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38672 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17327 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17378 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5434 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16971 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

GeorgesStavracas reviewed (approved):
LGTM

aperezdc commented on Source/WebCore/platform/graphics/skia/FloatRoundedRectSkia.cpp:
```diff
@@ -0,0 +1,64 @@
+    if (!isRounded())
+        return SkRRect::MakeRect(rect());
+
+    SkVector radii[4];
+    radii[SkRRect::kUpperLeft_Corner].set(m_radii.topLeft().width(), m_radii.topLeft().height());
+    radii[SkRRect::kUpperRight_Corner].set(m_radii.topRight().width(), m_radii.topRight().height());
+    radii[SkRRect::kLowerRight_Corner].set(m_radii.bottomRight().width(), m_radii.bottomRight().height());
+    radii[SkRRect::kLowerLeft_Corner].set(m_radii.bottomLeft().width(), m_radii.bottomLeft().height());
```
Any reason to not use an initializer expression? The code removed from `GraphicsContextSkia.cpp` did this:

```suggestion
    SkVector radii[4] = {
        {m_radii.topLeft().width(), m_radii.topLeft().height()},
        {m_radii.topRight().width(), m_radii.topRight().height()},
        {m_radii.bottomRight().width(), m_radii.bottomRight().height()},
        {m_radii.bottomLeft().width(), m_radii.bottomLeft().height()},
    };
```

GeorgesStavracas replied:
I'm not Carlos, but this opaque vector constructor was probably part of the reason why the borders are swapped and nobody noticed. I wonder if this would be an acceptable approach (using array indexing in the initializer):

```cpp
    SkVector radii[4] = {
        [SkRRect::kUpperLeft_Corner] = { m_radii.topLeft().width(), m_radii.topLeft().height() },
        [SkRRect::kUpperRight_Corner] = { m_radii.topRight().width(), m_radii.topRight().height() },
        [SkRRect::kLowerRight_Corner] = { m_radii.bottomRight().width(), m_radii.bottomRight().height() },
        [SkRRect::kLowerLeft_Corner] = { m_radii.bottomLeft().width(), m_radii.bottomLeft().height() },
    };
```

carlosgcampos replied:
Yes, I explained it in the other comment, with that initialization is easy to make a mistak because you need to know 0 is top left, 1 is to right, etc. And we actually made that mistake and corners are swapped. I prefer this way explicitly using the Corner enum values.

aperezdc replied:
Right, all fine then.

Also, I think the `{ [index] = value }` syntax is a compiler extension. Not sure if it's covered by the designated initializers feature but that would be C++20&mdash;and we are not yet requiring a C++20 compiler anyway.

carlosgcampos committed:
[Skia] Use Skia API when adding rounded rect to path and strategy is PreferNative
https://bugs.webkit.org/show_bug.cgi?id=269845

Reviewed by Adrian Perez de Castro.

Also add conversion operators between FloatRounrdedRect and SkRREct to
remove duplicated code.

* Source/WebCore/platform/SourcesSkia.txt:
* Source/WebCore/platform/graphics/FloatRoundedRect.h:
* Source/WebCore/platform/graphics/skia/FloatRoundedRectSkia.cpp: Added.
(WebCore::FloatRoundedRect::FloatRoundedRect):
(WebCore::FloatRoundedRect::operator SkRRect const):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::fillRoundedRectImpl):
(WebCore::GraphicsContextSkia::fillRectWithRoundedHole):
* Source/WebCore/platform/graphics/skia/PathSkia.cpp:
(WebCore::PathSkia::add):

Canonical link: https://commits.webkit.org/275106@main
